pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -464,6 +464,13 @@ func (og *operationGenerator) GenerateDetachVolumeFunc(
 			// On failure, add volume back to ReportAsAttached list
 			actualStateOfWorld.AddVolumeToReportAsAttached(
```

Comment:
I think, if we want 'uncertain' to take effect, we shouldn't add the volume-as-attached. If we do, then I think this is equivalent to adding the volume as Attached.

Reply:
good point. But I was concerning if volume is reported as not attached, but if detach does fail. In case of volume is required to be attached again, whether node status will be updated to reflect this.

So I checked again for this case. say, volume is still attached, but node status showing volume is not attached. If there is a new request for volume to be attached back to the same node, because volume is in uncertain state, attach operation will be triggered, after it succeeded, asw will call MarkVolumeAsAttached. This function will update volumesToReportAsAttached and mark node should be updated.  So basically it will make node updater to update this volume into attached list.

So far I think it is ok to remove this. 

btw: the main reason of this change is updated in the commit msg.

Reply:
Yes, for uncertain volumes:
- If the volume is NOT in DSW: The [asw.GetAttachedVolumes](https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L142) still return the volume, but the following dsw.VolumeExists() check fails. So, it will be detached.
- If the volume is in DSW: [asw.IsVolumeAttachedToNode](https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L229) fails. So, it will be attached.

And yes, for the fix to work as intended, we should remove logic that calls AddVolumeToReportAsAttached() here.

Reply:
there is a possible downside of removing the call to `AddVolumeToReportAsAttached` , which is - say detach failed from a node because volume was mounted and we removed the volume from attached state and hence is not reported as attached on the node. And then KCM crashed - we will rebuild the state by listing node objects and processing `node.Status.VolumesAttached` - which won't report the volume and then possibly volume will not be detached. 

BTW IIRC marking volume as "uncertain" here will prevent the volume from being reported into node's status. But if call `MarkAsUncertain` after `AddVolumeToReportAsAttached` - then does it still get reported?

Reply:
Discussed today. For CSI, VolumeAttachment will have deletion timestamp, but the object will not be fully deleted until CSI driver returns success from detach.